### ISSUE TITLE
Support hidden holdout evaluators to prevent benchmark overfitting

### ISSUE THREAD

**Issue by tan-yong-sheng** [davebcn87/pi-autoresearch#81]
Today, the agent optimizes whatever benchmark script it can read (e.g. `.auto/measure.sh`).
That's great for iterative optimization, but it also means the agent can optimize for the
benchmark itself rather than the underlying objective — analogous to training on the test set.

## Motivation
Suppose a project has:
```text
.auto/measure.sh           # optimization benchmark
.auto/measure-holdout.sh   # hidden evaluation
```

If the agent can also read `measure-holdout.sh`, it starts optimizing for both — at which point the holdout is no longer an independent check, it's just another target. This mirrors data leakage in ML, but shows up anywhere evaluation criteria are visible: Lighthouse scores, compiler optimizations, benchmark suites, codegen, search ranking, latency benchmarks, etc.

## Why a prompt rule alone won't cut it
My first instinct was "just instruct the agent not to read the holdout file." I don't think that actually works, and I'd rather flag that up front than propose it as a solution:

- An agent with normal filesystem access can `cat` / `grep` / `ls` its way into anything that isn't blocked at the OS or process-permission level. A system-prompt rule is a norm, not a boundary — nothing stops it being violated once, and that's all it takes.
- Long autonomous loops make this worse. `pi-autoresearch` already does compaction and context resets by design (see `autoresearch.md`/`autoresearch.jsonl` resumability). A "never open  this file" instruction is exactly the kind of soft constraint that can quietly drop out of context across a compaction boundary.
- Even without directly reading the file, an agent can infer holdout behavior indirectly — through logs, error messages, git history, or observable side effects of the holdout script (timing, files touched, output shape). A prompt rule doesn't stop inference from side  channels either.

So I don't think there's a fully airtight solution here — flagging that honestly rather than overselling this as "solved." But I think we can get meaningfully better than the status quo (fully visible benchmark) without needing perfection.

## Proposal
Introduce an optional **hidden holdout evaluator**, enforced at the tool/process level rather than relying on the agent's cooperation:

```text
.auto/measure.sh            # visible, optimization benchmark
.auto/measure-holdout.sh     # hidden evaluator, never exposed to the agent
```
or as a profile:
```text
run_experiment(profile="default")
run_experiment(profile="holdout")
```

Enforcement options, roughly strongest to weakest:
1. **OS-level permissions** — the holdout script is owned by a different user/group than the
   agent's shell runs as, so it's a real boundary rather than a convention.
2. **Process isolation** — the extension shells out to run the holdout evaluator in a separate
   process/sandbox the agent has no shell access to, and only the resulting metric crosses
   back into the agent's context.
3. **Convention + audit trail** (weakest, but cheap and useful as a fallback) — a
   `.gitignore`-style naming pattern plus a check that flags if any agent tool call ever
   touched the holdout path, so you at least get visibility into a leak rather than silent
   trust.

Possible run behavior regardless of enforcement tier:
- The agent receives only the resulting metric (or pass/fail), never the implementation.
- Holdout evaluation could run: only at the end, every N accepted experiments, or on demand.

## Benefits
- Reduces (though doesn't eliminate) benchmark overfitting.
- Better signal on whether improvements generalize, even if imperfect.
- Mirrors train/validation/test workflows without being ML-specific.
- Works for any optimization target where the benchmark itself should stay independent.

## Questions
- Would this fit the scope of `pi-autoresearch`, or is it better as a convention documented for users to implement themselves (e.g. via `.auto/hooks/`)?
- Is process isolation (option 2 above) worth the added complexity, or is an audited convention (option 3) good enough given no approach is airtight anyway?

This feels complementary to #69 (noise-aware repeats / confidence): that addresses *measurement noise*, this addresses *benchmark leakage* — and I want to be upfront that, unlike the noise problem, I don't think this one has a clean fix, just better and worse mitigations.

Just for discussions, thanks for this amazing tool.

**Comment by tan-yong-sheng** [comment]
Trying to mitigate this problems via autoresearch hooks to check the overfit ratio ... Just exploration, not sure if this is good way or not
